crates/sui-snapshot/src/restorer.rs

```diff
@@ -0,0 +1,196 @@
+        let remote_object_store = self.archive_reader.remote_object_store();
+        let cloned_state_sync_store = state_sync_store.clone();
+        let cloned_manifest = manifest.clone();
+        let cloned_remote_object_store = remote_object_store.clone();
+        let concurrency = self.archive_reader.concurrency();
+
+        let checkpoint_sync_handle = spawn_monitored_task!(async move {
+            load_summaries_upto(
+                epoch.saturating_add(1),
+                cloned_state_sync_store,
+                cloned_manifest,
+                concurrency,
+                cloned_remote_object_store,
+            )
+            .await
+            .expect("Failed to load summaries");
+        });
+
+        let (sha3_digests, accumulator, all_files) = self.snapshot_reader.get_checksums()?;
+        let (abort_handle, abort_registration) = AbortHandle::new_pair();
+        let cloned_snapshot_reader = self.snapshot_reader.clone();
+        let cloned_perpetual_db = perpetual_db.clone();
+
+        let read_handle = spawn_monitored_task!(async move {
+            cloned_snapshot_reader
+                .read(
+                    &cloned_perpetual_db,
+                    sha3_digests,
+                    all_files,
+                    abort_registration,
+                )
+                .await
+                .map_err(|e| anyhow::anyhow!("{:?}", e.to_string()))
+                .unwrap_or_else(|e| panic!("Failed to read snapshot from remote store: {:?}", e));
+        });
+
+        // Upon successful exit of this task, we should have all checkpoint summaries
+        // from genesis to the restore epoch, as well as the end of epoch checkpoint
+        // mapping populated.
+        checkpoint_sync_handle.await?;
+
+        if disable_verify {
+            warn!("Skipping formal snapshot verification! This is not recommended for production/non-emergency use.");
+        } else {
+            let checkpoint_summary = state_sync_store
+                .get_epoch_last_checkpoint(epoch)?
+                .expect("Expected last checkpoint for epoch to exist after summary sync");
+            let commitments = &checkpoint_summary
+                .end_of_epoch_data
+                .as_ref()
+                .unwrap_or_else(|| {
+                    panic!("Expected end of epoch checkpoint summary to be returned")
+                })
+                .epoch_commitments;
+            if commitments.is_empty() {
+                panic!(
+                    "Formal snapshot verification not supported for epoch {} on this node",
+                    epoch
+                );
+            }
+
+            let root_state_digest = ECMHLiveObjectSetDigest::from(accumulator.digest());
+            let CheckpointCommitment::ECMHLiveObjectSetDigest(commitment_digest) = &commitments[0];
+            if root_state_digest != commitment_digest.clone() {
+                abort_handle.abort();
+                panic!(
```

Comment:
I'm of the opinion that it would be better for the node operator to know that they failed to restore, given that the snapshot restore can take a long time. I imagine it would be worse to run a restore that fails for a correctible reason, and start syncing from genesis and require days of syncing to catch up, vs for it to fail and require intervention to either fix the issue or sync from genesis intentionally.

Also, this is integrated into a sui-tool command that then requires a step to start the node after the restore, so either way there is an additional step.

Reply:
I agree with your assessment that we should fail loudly if snapshot restore fails. My suggestion was if this panic could be instead moved under sui-tool command. I thought maybe one day we will have automatic snapshot restore in sui binary in which case we could perhaps have something like - if db is behind by `x` checkpoints, try formal snapshot restore but if that fails, try db snapshot restore or let the binary start as is (state-sync based restore).

Reply:
But for now if you want to leave the panic here, that is also fine.